qiskit_nature/second_q/operators/majorana_op.py

```diff
@@ -0,0 +1,521 @@
+        num_so = self.num_spin_orbitals
+        if other is not None:
+            other_num_so = other.num_spin_orbitals
+            if num_so is None:
+                num_so = other_num_so
+            elif other_num_so is not None:
+                num_so = max(num_so, other_num_so)
+
+        return self.__class__(data, copy=False, num_spin_orbitals=num_so)
+
+    def _validate_keys(self, keys: Collection[str]) -> None:
+        super()._validate_keys(keys)
+
+        num_so = self.num_spin_orbitals
+
+        max_index = -1
+
+        for key in keys:
+            # 0. explicitly allow the empty key
+            if key == "":
+                continue
+
+            # 1. validate overall key structure
+            if not re.fullmatch(MajoranaOp._OPERATION_REGEX, key):
+                raise QiskitNatureError(f"{key} is not a valid MajoranaOp label.")
+
+            # 2. validate all indices against register length
+            for term in key.split():
+                index = int(term[1:])
+                if num_so is None:
+                    if index > max_index:
+                        max_index = index
+                elif index >= 2 * num_so:
+                    raise QiskitNatureError(
+                        f"The index, {index}, from the label, {key}, exceeds twice the number of "
+                        f"spin orbitals, {num_so}."
+                    )
+
+        if num_so is None:
+            self.num_spin_orbitals = (max_index + 1 if max_index % 2 else max_index + 2) // 2
+
+    @staticmethod
+    def _majorana_label(label: str) -> str:
+        """Converts a Fermionic label into a Majorana label."""
+        return label.replace("+", "").replace("-", "")
+
+    @classmethod
+    def _validate_polynomial_tensor_key(cls, keys: Collection[str]) -> None:
+        # PolynomialTensor keys cannot be built from empty string,
+        # hence we choose _ to be the only allowed character
+        allowed_chars = {"_"}
+
+        for key in keys:
+            if set(key) - allowed_chars:
+                raise QiskitNatureError(
+                    f"The key {key} is invalid. PolynomialTensor keys may only consists of `_` "
+                    "characters, for them to be expandable into a MajoranaOp."
+                )
+
+    @classmethod
+    def from_polynomial_tensor(cls, tensor: PolynomialTensor) -> MajoranaOp:
+        cls._validate_polynomial_tensor_key(tensor.keys())
+
+        data: dict[str, _TCoeff] = {}
+
+        for key in tensor:
+            if key == "":
+                data[""] = tensor[key].item()
+                continue
+
+            mat = tensor[key]
+
+            empty_string_key = ["" for _ in key]  # label format for Majorana is just '_<index>'
+            label_template = mat.label_template.format(*empty_string_key)
+
+            for value, index in mat.coord_iter():
+                data[label_template.format(*index)] = value
+
+        num_so = (tensor.register_length + 1) // 2
```

Comment:
This is the same code as in my suggestion above but with the `+` and `// 2` operations reversed. Pick whichever one you like, but I guess we should be consistent to avoid confusion.